lib/streamlit/elements/alert.py

```diff
@@ -22,12 +22,12 @@
 
 
 class AlertMixin:
-    def error(self, body: str) -> "DeltaGenerator":
+    def error(self, body: object) -> "DeltaGenerator":
```

Comment:
Do you happen to have any examples of use-cases where one of these functions is given a non-`str` body?

While the `clean_text` function can take an arbitrary Python object as input, I can see an argument for us wanting to keep the tighter type restrictions unless passing objects to these functions is somewhat common

Reply:
Some arguments for _not_ loosening the restrictions, as I ended up doing:
- `str` better reflects the _intended_ usage than `object`, and might allow users to more quickly understand the purpose of the function. 
  > Aha! The function accepts a simple string, then in probably renders text.
- Avoid over-promising functionality that you might want to stop maintaining in the future. Say you want to remove the `clean_text` indirection, or something.

Now, here's why I ended up doing it anyway:

Me and my colleagues use Streamlit (and strict typing) extensively at work, so loosening was made based on our practical experiences. When I first annotated these functions, I didn't take our own code base into account, simply assuming that we always passed in a string to these methods, like the docstring suggests. Upon making use of a new build of Streamlit, `mypy` alerted me to the fact that we quite commonly pass in non-string objects to the alert methods. This was common enough that I thought it would be painful (and unnecessary) to add explicit string transformations in all cases. If this was true for us, I assumed it would be for other Streamlit API users as well.

The cases range from base types, like `int`s, to complex objects that have "nice" representations given applications of `clean_text`, like exceptions. I can give more details if needed. In an untyped setting, I think this "misuse" just naturally happens in cases where the _natural domain_ of a function or method is significantly wider than the _restricted domain_ implied by the docs. If the call does not result in an exception, and produces meaningful output, users don't care about the documentation frowning upon their use. After all, you usually only end up reading the docs properly when something goes wrong.

When adding annotations to a library that hasn't been annotated for a long time, there is a strong case to be made for annotating in such a way that you cover as large a subset of the _natural domain_ as practically possible. **This avoids prompting users to fix non-issues upon upgrading.** This is really my main motivation here.

Some thoughts on my initial counter-arguments:
- I did consider writing a `SupportsStr` protocol in order to annotate this function in a way that better hints at its intended use. But this essentially equates to `object` anyway, so I didn't feel it added enough value.
- Removing `clean_text` would be a significant breaking change anyway, also in the untyped setting, that would result in a major version bump. This isn't changed by loosening the typing.

Reply:
Update:

I ended up actually implementing the `SupportsStr` protocol, after thinking about it. I guess it gives some level of protection against types that have overridden `__str__`  to do unexpected things. Probably uncommon, but doesn't hurt to cover it.

I also reverted the docs to claim the input type is `str`, leading to a slight mismatch between the docstring and annotations, but I think that is an ok compromise to make, in order to make the docs more suggestive of the intended usecase, while avoiding unnecessary type check errors.

Reply:
Thanks for the explanation. I agree with the rationale for keeping the types loose given that it does seem like if you've personally realized that many of the calls in your codebase to `st.error`, etc, use various easily string-serialized types, it's likely that many codebases do the same. If there are enough codebases that both do this and use strict type-checking, then tightening the bounds would result in a breaking change that we'd want to avoid.

As for keeping the docs slightly "incorrect" given the actual type annotation, I'm also fairly okay with the mismatch to signal our intended usage of `st.error` and related elements.

The `SupportsStr` protocol seems like a good call to me given there's some level of protection that it offers, and at the very least the type can act as documentation even if it ends up being not too different from `object` in practice.